sentry_sdk/integrations/sanic.py

```diff
@@ -180,16 +194,43 @@ async def _hub_enter(request):
         scope.clear_breadcrumbs()
         scope.add_event_processor(_make_request_processor(weak_request))
 
+    transaction = continue_trace(
+        dict(request.headers),
+        op=OP.HTTP_SERVER,
+        # Unless the request results in a 404 error, the name and source will get overwritten in _set_transaction
+        name=request.path,
+        source=TRANSACTION_SOURCE_URL,
+    )
+    request.ctx._sentry_transaction = request.ctx._sentry_hub.start_transaction(
+        transaction
+    ).__enter__()
+
 
-async def _hub_exit(request, **_):
-    # type: (Request, **Any) -> None
-    request.ctx._sentry_hub.__exit__(None, None, None)
+async def _hub_exit(request, response):
+    # type: (Request, BaseHTTPResponse) -> None
+    with capture_internal_exceptions():
+        if not request.ctx._sentry_do_integration:
+            return
 
+        integration = Hub.current.get_integration(SanicIntegration)  # type: Integration
 
-async def _set_transaction(request, route, **kwargs):
+        # This capture_internal_exceptions block has been intentionally nested here, so that in case an exception
+        # happens while trying to end the transaction, we still attempt to exit the hub.
+        with capture_internal_exceptions():
+            request.ctx._sentry_transaction.set_http_status(response.status)
+            request.ctx._sentry_transaction.sampled &= (
+                isinstance(integration, SanicIntegration)
```

Comment:
Is this check necessary? Or is it there for type checking purposes?

Reply:
We could probably equivalently check that `integration` is not `None`, but this variation is more type-safe since we guarantee that `integration._unsampled_statuses` will not result in an `AttributeError` (and if we ever were to rename `_unsampled_statuses`, MyPy would warn us to update this reference as well).